rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -56,973 +51,869 @@ trait Reduce[M[_]] {
+                       } yield resultExp
+                     case (_: GInt, other) =>
+                       OperatorExpectedError("-", "Int", other.typ).raiseError[M, Expr]
+                     case (other, _) =>
+                       OperatorNotDefined("-", other.typ).raiseError[M, Expr]
+                   }
+        } yield result
 
-        case EMatchesBody(EMatches(target, pattern)) =>
-          for {
-            evaledTarget <- evalExpr(target)
-            substTarget  <- substituteAndCharge[Par, M](evaledTarget, 0, env)
-            substPattern <- substituteAndCharge[Par, M](pattern, 1, env)
-            matchResult  <- spatialMatchAndCharge[M](substTarget, substPattern)
-          } yield GBool(matchResult.isDefined)
-
-        case EPercentPercentBody(EPercentPercent(p1, p2)) =>
-          def evalToStringPair(keyExpr: Expr, valueExpr: Expr): M[(String, String)] =
-            (keyExpr.exprInstance, valueExpr.exprInstance) match {
-              case (GString(keyString), GString(valueString)) =>
-                Applicative[M].pure[(String, String)](keyString -> valueString)
-              case (GString(keyString), GInt(valueInt)) =>
-                Applicative[M].pure[(String, String)](keyString -> valueInt.toString)
-              case (GString(keyString), GBool(valueBool)) =>
-                Applicative[M].pure[(String, String)](keyString -> valueBool.toString)
-              case (GString(keyString), GUri(uri)) =>
-                Applicative[M].pure[(String, String)](keyString -> uri)
-              // TODO: Add cases for other ground terms as well? Maybe it would be better
-              // to implement cats.Show for all ground terms.
-              case (_: GString, value) =>
-                s.raiseError[(String, String)](
-                  ReduceError(s"Error: interpolation doesn't support ${value.typ}")
-                )
-              case _ =>
-                s.raiseError[(String, String)](
-                  ReduceError("Error: interpolation Map should only contain String keys")
-                )
-            }
-          @SuppressWarnings(
-            Array("org.wartremover.warts.Var", "org.wartremover.warts.NonUnitStatements")
-          )
-          // TODO consider replacing while loop with tailrec recursion
-          def interpolate(string: String, keyValuePairs: List[(String, String)]): String = {
-            val result  = StringBuilder.newBuilder
-            var current = string
-            while (current.nonEmpty) {
-              keyValuePairs.find {
-                case (k, _) => current.startsWith("${" + k + "}")
-              } match {
-                case Some((k, v)) =>
-                  result ++= v
-                  current = current.drop(k.length + 3)
-                case None =>
-                  result += current.head
-                  current = current.tail
-              }
+      case ELtBody(ELt(p1, p2)) => // TODO: Replace <* with <<
```

Comment:
I'm a bit partial on your leaving of TODO comments in a commit that does an unrelated thing. OFC, TODO are always unrelated to the thing you work on when you spot them, but it hurts even more when you leave the same TODO multiple times, and then get rid of all (but one) occurrences in the next commits.

At the same time, I see the workflow that emerges from such practice and am tempted to employ it myself. Can we though agree that - when there's multiple occurences of the same TODO, and the occurrences can be searched moderately easily - we don't include more than one copy of a TODO?

Reply:
Yes. We can agree :)